Remove duplicated IR ASTs for static methods

Akirathan opened the pull request:
Closes #11686

### Pull Request Description
Implements the newly specified behavior of method invocations from #13961.

### Important Notes

- IR cache size decreased by [almost 30%](https://github.com/enso-org/enso/pull/13962#issuecomment-3410756290).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Ensure IR cache size decreased. Update new expected value.
- [ ] Add links to test in https://github.com/enso-org/enso/pull/13961#discussion_r2397627368
- [ ] Remove static method wrappers for builtin methods (`staticOrInstanceMethod`)
- [x] The documentation has been updated, if necessary.
- [x] Unit tests have been written where possible.

Commit:
docs for dump: run with --no-ir-caches

Akirathan commented:
Looking into generated IR bindings for `Vector.to_array`:
<img width="1768" height="816" alt="image" src="https://github.com/user-attachments/assets/aaf64ea0-adfa-4b33-ab3f-cfaf5b1ff4ce" />
- There are two bindings ([Method.Explicit](https://github.com/enso-org/enso/blob/fd9460df4f98a01b08f720d2646ad14b5bdbae37/engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/module/scope/definition/Method.scala#L53) IR trees).
- The first one is created in [FunctionBinding](https://github.com/enso-org/enso/blob/fd9460df4f98a01b08f720d2646ad14b5bdbae37/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/desugar/FunctionBinding.scala#L184).
- The second one is created in [MethodDefinitions.generateStaticAliasMethod](https://github.com/enso-org/enso/blob/fd9460df4f98a01b08f720d2646ad14b5bdbae37/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/resolve/MethodDefinitions.java#L224-L239).
- The only difference between definition arguments 10 and 12 is:
  - 12 has `Name$SelfType` as `ascribedType`.
  - Note that 12 is on `arg[1]` position, not on `arg[0]`.

This PR aims to remove duplicated IR trees created in `MethodDefinitions` pass. In the above picture, this is the second tree with dashed edges.

Akirathan commented:
Will ensure that IR cache size was decreased after #14001 is merged.

Commit:
Tests for duplicated method in MethodDefinitions compiler pass

Commit:
Remove static method duplication from MethodDefinitions pass

Akirathan commented:
Removed duplicate static method registration from `MethodDefinitions` pass in 0af145d9059792a07b5e89006a94e709293acf31. Let's wait on the results of tests.

Commit:
ModuleScope does not have any eigen types registered.

- All the methods are registered on non-eigen types.
- Deprecate `Method.Explicit.isStatic` field.
- `Method.Explicit.isStaticWrapper` field will be removed

Commit:
Remove creation of singleton/eigen types.

- Every type is just a type.
- No more distinction between eigen and non-eigen type.

Commit:
Revert "Remove creation of singleton/eigen types."

This reverts commit 2a1152ec4dad6b5b71b50aa9e2dbb84e4108ee45.

Commit:
Revert "ModuleScope does not have any eigen types registered."

This reverts commit aaa4b061c1916835e3a7cb21679577bdec4b19d3.

Commit:
Test that method is registered just once in the scope.

Akirathan commented:
Testing methods registered on `ModuleScope`. More specifically, testing that a method is registered just once on the `ModuleScope`. In b9ad2d5d3a273a3643c18c9a3db33b2d30f2aec8

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/scope/ModuleScopeTest.java:
```diff
@@ -65,40 +47,162 @@ public void staticMethodIsInResolvedExports() throws IOException {
+    var assocType = mainRuntimeMod.getScope().getAssociatedType();
+    assertThat(assocType, is(notNullValue()));
+    var moduleMethod = mainRuntimeMod.getScope().getMethodForType(assocType, "module_method");
+    assertThat(moduleMethod, is(notNullValue()));
```
- I'd _assert function schema_ rather than just a simple `assertNotNull` check
- the schema assert would show us the `self` argument position
- that's an important information

Akirathan replied:
Good idea. Done in 3b07105e26ce71755bb2376d924089d57663b20b

JaroslavTulach commented on docs/runtime/compiler-ir.md:
```diff
@@ -29,6 +29,9 @@ enso --no-compile-dependencies --no-ir-caches --vm.D enso.compiler.RemoveUnusedI
 ## Dumping IR
 
+Note that to properly dump module from standard library, **it is recommended to
+run with --no-ir-caches**.
```
As this PR will modify the `.ir` and `.bindings` files significantly, it should [change the checksum header](https://github.com/enso-org/enso/blob/9390adb60ecdf69112b4a84b2d87e108f6adc3fb/docs/runtime/ir-caching.md#portability-and-versioning) to avoid mixing the new caches with those from other branches.

Akirathan replied:
Good idea. Done in 6cce54fbaea07754dd51eab315e4b62bedd86703

Commit:
Bump Persistance serial version

Commit:
Assert method schema.

Assert that there is only a single `self` argument

Commit:
Test that for all methods, there is just a single self argument

Commit:
Merge branch 'develop' into wip/akirathan/11686-register-only-static-methods

# Conflicts:
#	engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/scope/ModuleScopeTest.java
#	lib/java/persistance/src/main/java/org/enso/persist/PerMap.java

Commit:
fmt

Commit:
In generateReexportBindings, fix extension method lookup

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/scope/ModuleScopeTest.java:
```diff
@@ -137,4 +249,128 @@ public void importedStaticMethodIsRegisteredInModuleScope() throws IOException {
+    }
+  }
+
+  private static void assertOnlyFirstArgumentIsSelf(Function function) {
```
I am fan of helper `assertSomething` methods.

Commit:
Special function lookup for self=Any.

There are no registered methods on Any.type eigen type.

Commit:
Extension methods are first looked up on eigen type

Commit:
Remove self arg in special circumstances from method call.

Commit:
Try to resolve function first on eigen type, then on normal type.

Commit:
Remove imprecise assert

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -155,15 +158,20 @@ Object doFunctionalDispatchCachedSymbol(
+      UnresolvedSymbol symbol, Object self, Type selfTpe, MethodResolverNode methodResolverNode) {
     Function function = methodResolverNode.executeResolution(selfTpe, symbol);
+    if (function == null && selfTpe.isEigenType()) {
+      // Try to resolve one more type, this time, not on eigen type, but on normal type.
```
```suggestion
      // Try to resolve one more time, this time, not on eigen type, but on normal type.
```

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -155,15 +158,20 @@ Object doFunctionalDispatchCachedSymbol(
+    if (function == null && selfTpe.isEigenType()) {
+      // Try to resolve one more type, this time, not on eigen type, but on normal type.
+      assert self instanceof Type;
+      function = methodResolverNode.executeResolution((Type) self, symbol);
```
Do you want to handle the _"fallback to static syntax for invoking instance"_ methods here?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -246,55 +250,52 @@ Object doFunctionalDispatchUncachedSymbol(
+  /**
+   * Special case of static method invocation on {@code Any}. More specifically, when the receiver
+   * ({@code self}) is {@code Any} type. In this situation, the self argument may be removed from
+   * the function call.
```
- this isn't how I'd like to see things implemented
- invoking a method sets `self=<receiver>` as _defaulted_ argument
   - e.g. `Integer.to_text` gets resolved to `Integer.to_text self=Integer` (`self` is defaulted)
   - and `Any.to_text` gets resolved to `Any.to_text self=Any` (`self` is defaulted)
- **static method call on Any** then means to explicitly change the `self`
   - e.g. `Any.to_text self="Hi"`
- such a syntax however works for all types - e.g. `Integer.to_text self=42`
- _no special handling_ of static dispatch of methods on `Any` shall be needed

JaroslavTulach replied:
The [idea with defaulted self](https://github.com/enso-org/enso/pull/13962#pullrequestreview-3287447276) is flawed, as first position argument in `func x=1 y = x+y` is applied to `x` not `y` - e.g. `func 42` leads to `tmp.func[tmp.enso] self=tmp x=42 y=_`. Thus the first position argument in case of defaulted `self` would go to the `self` itself, alas!

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala:
```diff
@@ -1035,18 +1038,24 @@ private[runtime] class IrToTruffle(
                       staticMethod.methodName
                     )
+                  if (fun == null) {
+                    fun = currentScope.getMethodForType(
```
- I don't like the `shouldRemoveSelfArg` method. Why is it needed?
- Is it needed only because here:
  - methods from `eigenTp` and `tp` are **mixed together** here?
- Don't mix those methods!
- Then you can avoid `shouldRemoveSelfArg`, I believe.

JaroslavTulach reviewed (commented):
- I have doubts about `shouldRemoveSelfArg` and `dispatchStaticMethodCallOnAny`
- we want to _remove these special casings_

### ~~Algorithm~~ Rejected!

**Update**: this algorithm is flawed, as first possition argument in `func x=1 y = x+y` is applied to `x` not `y` - e.g. `func 42` leads to `tmp.func[tmp.enso] self=tmp x=42 y=_`. Thus the first position argument in case of _defaulted_ `self` would go to the `self` itself, alas!

- invoking a method is a matter of two things:
   - resolve the method in the **receiver's type** (and provide _defaulted_ `self`)
   - invoke the method (possibly with explicit `self` argument)
- if no method is found in the receiver's type:
   - only then consider _static invocation syntax_
   - resolve the method in the **receiver itself** 
   - don't bind `self` to anything
   - invoke the method (with the first argument being `self` _non-defaulted_)

Commit:
Update Any_Spec to the new spec

JaroslavTulach commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,78 @@ add_specs suite_builder =
+            Test.expect_panic_with (Any.== Boolean) No_Such_Method
             (Any.== self=Boolean) . should_be_a Function
             (Any.== self=Any Boolean) . should_be_false
-            (Any.== Any Any) . should_be_true
-            (Any.== Boolean) . should_be_false
-            (Any.== Any) . should_be_true
-            (Any == Boolean) . should_be_false
-            (Any == Any) . should_be_true
```
- Don't delete these lines!
- `Any.==` is really a function because `==` is an instance method on `Any`

Akirathan replied:
Right. That was an accident. `Any == Any` is an instance method invocation, equivalent to `Any.== self=Any Any`. Added back in 6b17fd9248d8bfe04615f2aa28d29fa6b4b8014b

JaroslavTulach commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,78 @@ add_specs suite_builder =
-            (Any.== Boolean) . should_be_false
-            (Any.== Any) . should_be_true
-            (Any == Boolean) . should_be_false
-            (Any == Any) . should_be_true
+            (Any.== self=Any Any) . should_be_true
 
+        group_builder.specify "Static method invocation (to_text)" <|
             # Any.to_text is a method that takes one argument (self)
-            Any.to_text . should_equal "Any"
+            Test.expect_panic_with (Any.to_text) No_Such_Method
```
No, `Any.to_text` should equal "Any"

Akirathan replied:
Are you sure? `Any.to_text`:
- Is an [Instance method invocation](https://github.com/enso-org/enso/pull/13961/files#diff-20577053b7072d45d6e099b15636ad10993933479c642dec0ac43a6f69118212R231)
- Type of the receiver is `Any.type`
- There is no `Any.type.to_text` method.
  - i.e., no `to_text` method on `Any` eigen type.
  - Note that there is only `Any.to_text` method

JaroslavTulach replied:
It should be the case as we agreed at https://github.com/enso-org/enso/pull/13961/files#r2401345452 - you are correct `allTypes(Any.type)` has to be changed to return `[Any.type, Any]`.

Akirathan replied:
Updated to the newest specification in 7b589378a49dbcb104f40166c064526a31ece732

JaroslavTulach commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,78 @@ add_specs suite_builder =
             Any.to_text self=Any . should_equal "Any"
             Any.to_text self=Vector . should_equal "Vector"
             Any.to_text self=[1,2,3] . should_equal "[1, 2, 3]"
-            Any.to_text Any . should_equal "Any"
```
Changing these lines is correct. `Any.to_text` evaluates to `"Any"` and `("Any" Any)` yields _non-invocable error_.

Akirathan replied:
Updated in 7b589378a49dbcb104f40166c064526a31ece732

Commit:
Add back some tests

JaroslavTulach commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,78 @@ add_specs suite_builder =
-            Vector.contains [1,2,3] 1 . should_be_true
 
             # Vector.is_empty has just one argument (self)
-            Vector.is_empty . to_text . contains "Vector.type.is_empty" . should_be_true
```
Oh, no! This is bad! `Vector.is_empty` is supposed to work. Or at least we need some syntax how to get a reference to `is_empty` functions of `Vector`!

- functionally-wise one could `Vector.is_empty self=_`, but that creates a wrapper lambda function around it
- special syntax `Vector.is_empty self=` - that's a bad idea I assume
- possibly we can detect `Vector.is_empty self=_` while processing IR and turn it into a direct reference to `is_empty` method of `Vector` ....

Like walking in a vicious circle!

Akirathan replied:
We concluded that this should not be a problem. We will just create wrapper lambda functions, if necessary. References to instance methods should not be too frequent in our codebase anyway.

JaroslavTulach commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,78 @@ add_specs suite_builder =
 
+        group_builder.specify "Static method invocation (spec_method)" <|
             # Any.spec_method is a method that takes three arguments (self, arg1, arg2)
-            (Any.spec_method) . should_be_a Function
-            (Any.spec_method Boolean) . should_be_a Function
```
As the `spec_method` is an _instance extension_ method of `Any`:
```
Any.spec_method self arg1 arg2 =
    "Any.spec_method:" + "{" + self.to_text + "}{" + arg1.to_text + "}{" + arg2.to_text + "}"
```
then both `Any.spec_method` and `(Any.spec_method Boolean)` are valid, I think.

Akirathan replied:
Yes, they are. Updated in 7b589378a49dbcb104f40166c064526a31ece732

Commit:
Update Any_Spec to the newest specification.

Newest specification according to https://github.com/enso-org/enso/pull/13961/commits/ed299f5be5bc3a4081bbd63a5268a375820ea920

JaroslavTulach commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,79 @@ add_specs suite_builder =
 
     suite_builder.group "Any's methods" group_builder->
-        group_builder.specify "should not be overridable when called statically" <|
+        group_builder.specify "Instance method invocation" <|
```
This diff is more important than https://github.com/enso-org/enso/pull/13961#issuecomment-3366485097 and it’s https://github.com/enso-org/enso/commit/7b589378a49dbcb104f40166c064526a31ece732 as it shows the diff against current behavior. Can we make it *less red*?

Green lines (additions) are OK. Red lines (removals) need deeper review. Hence the request to make it *less red*.

Akirathan replied:
Done in https://github.com/enso-org/enso/commit/52efe5f9bd87e31679f866c9e7ab80fe344abce1, and added some comments. I think now it is pretty reviewable.

Commit:
Make tests diff less red

Akirathan commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,89 @@ add_specs suite_builder =
             Any.spec_method Boolean Vector . should_equal "Any.spec_method:{Any}{Boolean}{Vector}"
-            Any.spec_method Any Boolean Vector . should_equal "Any.spec_method:{Any}{Boolean}{Vector}"
-            Any.spec_method Date Boolean Vector . should_equal "Any.spec_method:{Date}{Boolean}{Vector}"
+            Test.expect_panic_with (Any.spec_method Any Boolean Vector) Not_Invokable
```
Not_Invokable, because `Any.spec_method Any Boolean Vector` is equivalent to `Any.spec_method self=Any Any Boolean Vector`.

Akirathan commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,89 @@ add_specs suite_builder =
             # Vector.contains has two arguments (self elem)
-            Vector.contains . to_text . contains "Vector.type.contains" . should_be_true
-            (Vector.contains [42]) . to_text . contains "Vector.type.contains" . should_be_true
+            Test.expect_panic_with (Vector.contains) No_Such_Method
```
No_Such_Method because there is no `contains` method in `Vector.type, neither in `Any`.

Akirathan commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,89 @@ add_specs suite_builder =
-            Any.to_text Any . should_equal "Any"
-            Any.to_text Vector . should_equal "Vector"
-            Any.to_text [1,2,3] . should_equal "[1, 2, 3]"
+            Test.expect_panic_with (Any.to_text Any) Not_Invokable
```
Not_Invokable because `Any.to_text Any` is equivalent to `Any.to_text self=Any Any` and thus evaluates to `"Any" Any`

Akirathan commented on test/Base_Tests/src/Semantic/Any_Spec.enso:
```diff
@@ -38,75 +38,89 @@ add_specs suite_builder =
             My_Type.x . should_equal "Any:My_Type"
             With_X.x . to_text . should_equal "Any:With_X"
-            With_X.y.should_be_a Function
+            Test.expect_panic_with (With_X.y) No_Such_Method
```
No_Such_Method, because there is no `y` method in `With_X.type` eigen type and neither in `Any`.

Commit:
Add tests for singleton type

Commit:
Merge branch 'develop' into wip/akirathan/11686-register-only-static-methods

Commit:
[WIP] Skeleton of StaticInvokeMethodNode

Commit:
Add TypeChainTest from #13961

Commit:
Implement correct semantics of Type.allTypes

Commit:
Merge branch 'develop' into wip/akirathan/11686-register-only-static-methods

# Conflicts:
#	engine/runtime-compiler/src/main/java/org/enso/compiler/pass/resolve/MethodDefinitions.java
#	engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeChainTest.java
#	lib/java/persistance/src/main/java/org/enso/persist/PerMap.java

Commit:
Add errorChain test

Commit:
fmt

Commit:
refact

Commit:
doc lin k

Commit:
invokeFuncNodce field

Commit:
Implement simple static method invocation

Commit:
Try to resolve method also on eigen type

Commit:
fix imports in Any_Spec

Commit:
remove assert: call schema can be empty

Commit:
Do not resolve method on eigen type in InstanceMethodInvokeNode

Commit:
Add ModuleScopeTest for instance method on singleton type

Commit:
Add TypeChainTest for eigen type from normal type

Commit:
Do not resolve on eigen type in InvokeMethodNode.resolveFunction

Commit:
Update tests in ModuleScopeTest

Commit:
fmt

Commit:
No special resolution for "static": methods overriden from Any

Commit:
Add MethodResolutionTest for resolving extension static method overridden from Any

Akirathan commented:
So far, the IR cache size was decreased by almost **30 MB**:
- On develop: [117.86 MB](https://github.com/enso-org/enso/actions/runs/18541356115/job/52849433277?pr=14135#step:10:3440)
- On this PR: [88.21 MB](https://github.com/enso-org/enso/actions/runs/18535896275/job/52830844450?pr=13962#step:10:3116)

Commit:
Fix getting static method reference to binary operator

Commit:
Fix static method invocations in Date

Commit:
More static method invocations fixes in tests

Commit:
revert Type.InvokeMember to develop:

self argument can potentially be prepended to the call

Commit:
StaticInvokeMethodNode might not remove first arg

Commit:
update test: Ref.get is an instance method

Commit:
Type.InvokeMember: For builtin type, try to resolve the method both on standard type and on eigen type

Commit:
fmt

Commit:
Add SelfParameterTest.

Testing implicit self parameter.

Commit:
Replace `Any.== x y` with `x == y`.

There is no static method invocation for binary dispatched methods.

Commit:
Merge branch 'develop' into wip/akirathan/11686-register-only-static-methods

Commit:
Fix argument for StaticInvokeMethodNode ctor

Commit:
Test that conversion is applied to self argument

Commit:
Add type ascription to self parameter

Commit:
fmt

Commit:
Remove unused argument

Commit:
Synthetic self parameter has eigen type as ascribed type

Commit:
Remove commented code

Commit:
Methods from Comparable are invoked statically

Commit:
Error.to_text is called statically

Commit:
Integer.+ is not a static method

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/AddNode.java:
```diff
@@ -11,7 +11,7 @@
 
   @Override
-  abstract Object executeBinary(Object own, Object that);
+  abstract Object executeBinary(Object self, Object that);
```
On develop, the fact that a `@BuiltinMethod` has first parameter with a name different than `self` means that there will be `AddMethodGen` generated with 3 parameters: [`self`, `own`, `that`]. More specifically, there will be two `Function` objects created. One with 3 parameters, and another one with 4 parameters (its static variant): [`selfStatic`, `self`, `own`, `that`].

Commit:
Revert "Integer.+ is not a static method"

This reverts commit ec2cb19d488f7b87d6ecf0290e6de1db3122c1c2.

EDIT: The builtin `integer_plus` **should** be static

Akirathan commented:
## Binary dispatch based on `that` argument inconsistency

I believe I have found an inconsistency in binary dispatch algorithm introduced in #8779. The inconsistency is manifested when a binary operator is called statically on develop (specified in [Numbers_Spec.enso](https://github.com/enso-org/enso/blob/2eb1bcd475680eef60c01f10b7f41b4aa989fb04/test/Base_Tests/src/Data/Numbers_Spec.enso#L238-L243)):
```
Integer.+ 1 0.5  . should_fail_with Type_Error
1 + 0.5  . should_equal 1.5
```
In other words, why `Integer.+ 1 0.5` is not semantically equivalent to `1 + 0.5`?

`Integer.+ 1 0.5`:
- Is calling `Integer.type.+ self=Integer.type self=1 that=0.5` static method wrapper of an instance method
- Which, in turn, calls `Integer.+ self=1 that=0.5` instance method.
- And that fails on Type_Error, because `0.5` cannot be converted to `Integer`.

`1 + 0.5`:
- Is correctly recognized as a binary operator, so binary dispatch will be tried.
- First dispatch based on `self` fails:
  - `Integer.+ self that:Integer` method is found.
  - `0.5` cannot be converted to `Integer`.
- Second dispatch based on `that` succeeds:
  - `Float.+ self that:Float` method is found.
  - `1` is converted to `Float`.	

--------------

**With this PR, such inconsistency is removed**, because:
- `Integer.+ 1 0.5` results in No_Such_Method panic
- `Integer.+ self=1 0.5` results in 1.5.
- `1 + 0.5` has the same behavior as in develop.

Commit:
update Number_Spec test: binary dispatch incosistency

Akirathan commented on test/Base_Tests/src/Data/Numbers_Spec.enso:
```diff
@@ -236,11 +236,18 @@ add_specs suite_builder =
+            # `Integer.+` in this case is resolved to the builtin method `Integer.type.+`
             Integer.+ 1 2 . should_equal 3
-            Test.expect_panic_with (Integer.+ 1 2.5)  Type_Error
+            Integer.+ 1 2.5 . should_equal 3.5
```
Why is this working now? Because `Integer.+ 1 2.5` is resolved to `Integer.type.+ self=Integer own=1 that=2.5` which is a builtin method call. This builtin is handled in [AddNode](https://github.com/enso-org/enso/blob/63646c94c18d62613f25145b5deb38252872078a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/AddNode.java), which has a [specialization](https://github.com/enso-org/enso/blob/63646c94c18d62613f25145b5deb38252872078a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/integer/AddNode.java#L30-L33) when `that` is `double`.

Commit:
update expected static methods in Meta_Spec

Commit:
Interop invocation does not have to receive two self arguments

Commit:
Any.== return self if self is dataflow error

Commit:
self argument is type checked - fix ObjectComparator std-bits

Commit:
Fix InstanceInvokeNode.doWarning.

Warning type must be prepended as self argument.

Commit:
Fix static method calls in Warnings_Spec.

Commit:
Fix static method calls in Private_Spec.

Commit:
InstanceInvokeMethodNode - try to find method only on Error type

JaroslavTulach commented:
There  are:
- [20 failing tests](https://github.com/enso-org/enso/actions/runs/18781287183/job/53589737140?pr=13962#step:14:1911) mostly `Compable undefined symbol <`

JaroslavTulach committed:
Use the static invocation syntax with self

JaroslavTulach committed:
Additional self argument for call to hash_callback needed

JaroslavTulach committed:
Returning back the original submodule

Commit:
SortVectorNode respects preapplied self argument when sorting generic vector

Commit:
update MethodDefinitionsTests.

No method duplication.

Commit:
update static method invocation in AnyOrStaticTest

Commit:
Fix static method invocation on Comparable

Commit:
update BinaryDispatchTest: b inary dispatch works even for static method invocation

Commit:
TypeInference does not work for static method invocation.

Because of the named self argument - type inference does not work for named arguments

Commit:
TypeInference handles static method invocation

Commit:
IGVDumper: argument name is a separate node

Commit:
Implement StaticInvokeMethodNode.setId

Commit:
Merge branch 'develop' into wip/akirathan/11686-register-only-static-methods

# Conflicts:
#	lib/java/persistance/src/main/java/org/enso/persist/PerMap.java

Commit:
Fix NPE in SortVectorNode.getPreappliedSelfArgument

Commit:
update TypeInferenceTest: static method call does not have to attach type inference metadata

Commit:
Fix more static method invocations

Commit:
fmt

Commit:
fix indexes in RuntimeServerTest

Commit:
Two self args in method call are compilation error

Commit:
Fix the spans in RuntimeServerTest

Commit:
Merge branch 'develop' into wip/akirathan/11686-register-only-static-methods

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Ordering_Helpers.enso:
```diff
@@ -25,7 +25,7 @@ type Default_Comparator
     compare : Any -> Any -> (Ordering|Nothing)
     compare x y =
-        case Any.== x y of
+        case x == y of
```
Maybe rather 
```suggestion
        case Any.== self=x that=y of
```
I believe I must have added this strange static invocation for a reason. Maybe it was related to comparing `Error`s?

But if things work for you, then the simple `x == y` is nicer!

Akirathan replied:
> I believe I must have added this strange static invocation for a reason. Maybe it was related to comparing Errors?

I thought that as well. I believe `x == y` was causing some infinite recursion in some cases. But since tests are green now, I would just leave the simple `x == y` here.

Note that `Any.== self=x that=y` is not the same as `x == y`, because it explicitly binds `self` and `that` arguments. At least that is not the case on develop.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsBuiltinNode.java:
```diff
@@ -46,6 +47,9 @@ static EqualsBuiltinNode build() {
    * @return {@code true} if {@code self} and {@code that} seem equal
    */
   public Object execute(VirtualFrame frame, Object self, Object other) {
+    if (self instanceof DataflowError e) {
```
I see. That allows you to write `x == y` and not `Any.== x y`. Good.

JaroslavTulach commented on test/Base_Tests/src/Semantic/Meta_Spec.enso:
```diff
@@ -432,7 +432,7 @@ add_specs suite_builder =
 
         group_builder.specify "static methods of Integer" <|
-            expected_methods = ['%', '*', '+', '-', '/', '<', '<=', '>', '>=', '^', 'abs', 'bit_and', 'bit_not', 'bit_or', 'bit_shift', 'bit_shift_l', 'bit_shift_r', 'bit_xor', 'ceil', 'div', 'fits_in_long', 'floor', 'negate', 'parse', 'round', 'to_decimal', 'to_float', 'truncate']
+            expected_methods = ['parse']
```
Looks we are heading the right direction!

JaroslavTulach committed:
Use static invocation syntax

JaroslavTulach commented on test/Helpers/src/Main.enso:
```diff
@@ -13,7 +13,7 @@ type Public_Type
     create data = Priv_Type.create data
     in_ctx ~action = Priv_Type.in_ctx action...
-    private priv_method self = Priv_Type.priv_method
+    private priv_method self = Priv_Type.priv_method self=self
```
This should fix the Enso Linting step

JaroslavTulach reviewed (approved):
I think we are getting close to the integration!

JaroslavTulach committed:
Prepend self argument in case of static invocation

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypePropagation.java:
```diff
@@ -439,7 +443,12 @@ private TypeRepresentation processUnresolvedSymbolApplication(
           }
-          return resolvedStaticMethod;
+          if (isStaticMethodInvocation) {
+            var withSelf = TypeRepresentation.buildStaticMethod(typeObject, resolvedStaticMethod);
```
- there is failing [Test org.enso.compiler.test.TypeInferenceTest.staticCallWithWrongType](https://github.com/enso-org/enso/actions/runs/18991427221/job/54245455267#step:14:6273) test
- I think it should be fixed by 6b5f1917c972c410f449614678bca8f1402890b5
- which prepends the "self type" as an argument in case of _static invocation_

JaroslavTulach committed:
Reordering static methods in the generated API

Akirathan commented on engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/semantic/MethodsTest.scala:
```diff
@@ -180,14 +180,12 @@ class MethodsTest extends InterpreterTest {
           |main =
-          |    IO.println (Foo.inc (Foo.Mk_Foo 12))
           |    IO.println (Foo.Mk_Foo 13).inc
-          |    IO.println (.inc self=Foo self=(Foo.Mk_Foo 14))
```
Method call with two `self` arguments is now a compile error thanks to [NoDuplicateSelfInMethodCalls](https://github.com/enso-org/enso/blob/5f05e888ccdda9eac99e1f71e03cd6134ffe237e/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/NoDuplicateSelfInMethodCalls.java) compiler pass.
